singer/src/main/java/com/pinterest/singer/writer/KafkaWriter.java

```diff
@@ -254,18 +401,44 @@ public void writeLogMessages(List<LogMessage> logMessages) throws LogStreamWrite
 
       int bytesWritten = 0;
       int maxKafkaBatchWriteLatency = 0;
+      boolean anyBucketSendFailed = false;
+
       for (Future<KafkaWritingTaskResult> f : resultSet) {
         KafkaWritingTaskResult result = f.get();
         if (!result.success) {
           LOG.error("Failed to write messages to kafka topic {}", topic, result.exception);
-          throw new LogStreamWriterException("Failed to write messages to kafka");
+          anyBucketSendFailed = true;
         } else {
           bytesWritten += result.getWrittenBytesSize();
           // get the max write latency
           maxKafkaBatchWriteLatency = Math.max(maxKafkaBatchWriteLatency,
               result.getKafkaBatchWriteLatencyInMillis());
+          if (isLoggingAuditEnabledAndConfigured()) {
+            int bucketIndex = result.getPartition();
+            // when result.success is true, the number of recordMetadata SHOULD be the same as
+            // the number of ProducerRecord. The size mismatch should never happen.
+            // Adding this if-check is just an additional verification to make sure the size match
+            // and the audit events sent out is indeed corresponding to those log messages that
+            // are audited.
+            if (bucketIndex >= 0 && result.getRecordMetadataList().size() != buckets.get(bucketIndex).size()){
+              LOG.warn("Number of ProducerRecord does not match the number of RecordMetadata, "
+                  + "LogName:{}, Topic:{}, BucketIndex:{}, result_size:{}, bucket_size:{}",
+                  logName, topic, bucketIndex, result.getRecordMetadataList().size(),
+                  buckets.get(bucketIndex).size());
+              OpenTsdbMetricConverter.incr(SingerMetrics.AUDIT_HEADERS_METADATA_COUNT_MISMATCH, 1,
+                  "topic=" + topic, "host=" + HOSTNAME, "logStreamName=" + logName, "partition=" + bucketIndex);
+            } else {
+              enqueueLoggingAuditEvents(result, mapOfHeadersMap.get(bucketIndex));
```

Comment:
Please add a comment that this is the regular / happy path

Reply:
added comments in the update of the PR.